### PR TITLE
Add DataRecipe

### PR DESCRIPTION
**Pull request by tchaton**
## What does this PR do?

This PR introduces DataRecipe to help processing data in the cloud with more ease. Here is how a user could resize ImageNet.

```python
import os
from PIL import Image
from lightning.data import DataProcessor, DataTransformRecipe

class ImageNetResizeRecipe(DataTransformRecipe):
    def prepare_structure(self, input_dir):
        return [os.path.join(input_dir, filename) for filename in os.listdir(input_dir)]

    def prepare_item(self, output_dir, filepath):
        Image.open(filepath).resize((256, 256)).save(os.path.join(output_dir, os.path.basename(filepath)))

data_processor = DataProcessor(
    name="imagenet_resized/train",
    num_workers=32,
)
data_processor.run(ImageNetResizeRecipe())
```

Or the simplified version using map:

```python
import os
from PIL import Image
from lightning.data import map

def fn(output_dir, filepath):
    img = Image.open(filepath)
    img = img.resize((256, 256))
    img.save(os.path.join(output_dir, os.path.basename(filepath)))

map(fn, inputs)
```

Fixes #\<issue_number>

<details>
  <summary><b>Before submitting</b></summary>

- Was this **discussed/agreed** via a GitHub issue? (not for typos and docs)
- [ ] Did you read the [contributor guideline](https://github.com/Lightning-AI/lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [ ] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- Did you make sure to **update the documentation** with your changes? (if necessary)
- Did you write any **new necessary tests**? (not for typos and docs)
- [ ] Did you verify new and **existing tests pass** locally with your changes?
- Did you list all the **breaking changes** introduced by this pull request?
- Did you **update the CHANGELOG**? (not for typos, docs, test updates, or minor internal changes/refactors)

</details>

## PR review

Anyone in the community is welcome to review the PR.
Before you start reviewing, make sure you have read the [review guidelines](https://github.com/Lightning-AI/lightning/wiki/Review-guidelines). In short, see the following bullet-list:

<details>
  <summary>Reviewer checklist</summary>

- [ ] Is this pull request ready for review? (if not, please submit in draft mode)
- [ ] Check that all items from **Before submitting** are resolved
- [ ] Make sure the title is self-explanatory and the description concisely explains the PR
- [ ] Add labels and milestones (and optionally projects) to the PR so it can be classified

</details>

----
:books: Documentation preview :books:: https://pytorch-lightning--18892.org.readthedocs.build/en/18892/

cc @borda

### CODE REVIEW

**Review by justusschock — APPROVED** [review]

File: src/lightning/data/__init__.py [hunk 1]
```diff
@@ -1,12 +1,10 @@
 from lightning.data.datasets import LightningDataset, LightningIterableDataset
-from lightning.data.streaming.dataloader import StreamingDataLoader
 from lightning.data.streaming.dataset import StreamingDataset
-from lightning.data.streaming.dataset_optimizer import DatasetOptimizer
+from lightning.data.streaming.map import map
```
Comment (Borda) [inline src/lightning/data/__init__.py:3]: please do not use the same name as for building fn
https://www.geeksforgeeks.org/python-map-function/

File: src/lightning/data/streaming/__init__.py [hunk 2]
```diff
@@ -12,7 +12,7 @@
 # limitations under the License.
 
 from lightning.data.streaming.cache import Cache
+from lightning.data.streaming.data_processor import DataProcessor
 from lightning.data.streaming.dataloader import StreamingDataLoader
-from lightning.data.streaming.dataset_optimizer import DatasetOptimizer
 
-__all__ = ["Cache", "DatasetOptimizer", "StreamingDataLoader"]
+__all__ = ["Cache", "DataProcessor", "StreamingDataLoader"]
```
[the remark this answers is not in the document]
Comment (tchaton) [reply src/lightning/data/streaming/__init__.py:18]: Sure.

File: src/lightning/data/streaming/writer.py [hunk 3]
```diff
@@ -205,7 +205,9 @@ def _create_chunk(self, filename: str, on_done: bool = False) -> bytes:
             max_index = self._max_index
 
         if len(items) == 0:
-            raise RuntimeError("The items shouldn't have an empty length. Something went wrong.")
+            raise RuntimeError(
+                "The items shouldn't have an empty length. Something went wrong." f" Found {self._serialized_items}"
```
[the remark this answers is not in the document]
Comment (tchaton) [reply src/lightning/data/streaming/writer.py:209]: Done !

File: src/lightning/data/streaming/data_processor.py [hunk 4]
```diff
@@ -437,188 +423,297 @@ def _start_remover(self) -> None:
...
+                        chunk_filepath = self.cache._add_item(self._index_counter, item_data)
+                        self._try_upload(chunk_filepath)
+                        self._index_counter += 1
+            elif item_data_or_generator is not None:
+                chunk_filepath = self.cache._add_item(self._index_counter, item_data_or_generator)
+                self._try_upload(chunk_filepath)
+                self._index_counter += 1
+        except Exception as e:
+            print(f"Failed processing {self._current_item}")
+            raise e
```
[the remark this answers is not in the document]
Comment (tchaton) [reply src/lightning/data/streaming/data_processor.py:472]: Done !

File: src/lightning/data/__init__.py (hunk 1 above)
Comment (tchaton) [reply src/lightning/data/__init__.py:3]: This is coming from @williamFalcon

**Pull request merged**